File: sidecar/raylab_sidecar/diagnostics.py
from __future__ import annotations

import ipaddress
import os
import platform
import shutil
import socket
import subprocess
from dataclasses import dataclass

import psutil

from .bootstrap import PINNED_PYTHON, PINNED_RAY_VERSION, python_version as bootstrap_python_version, ray_version as bootstrap_ray_version, resolved_ray_executable
from .models import AppConfig, DiagnosticCheck


PRIVATE_LOOPBACKS = {"127.0.0.1", "localhost"}


def is_private_host(host: str) -> bool:
    if host in PRIVATE_LOOPBACKS:
        return True
    try:
        ip = ipaddress.ip_address(socket.gethostbyname(host))
    except OSError:
        return False
    return ip.is_private or ip.is_loopback


def is_port_available(host: str, port: int) -> bool:
    try:
        bind_host = socket.gethostbyname(host)
    except OSError:
        bind_host = "127.0.0.1"
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        try:
            sock.bind((bind_host, port))
        except OSError:
            return False
    return True


def is_port_reachable(host: str, port: int, timeout: float = 1.0) -> bool:
    try:
        with socket.create_connection((host, port), timeout=timeout):
            return True
    except OSError:
        return False


def is_coordinator_reachable(config: AppConfig, timeout: float = 1.0) -> bool:
    try:
        from .discovery import _probe_user, _tcp_open

        return _tcp_open(
            config.coordinator.head_host,
            config.coordinator.ray_port,
            timeout,
            allow_cli_fallback=True,
            cli_user=_probe_user(config),
        )
    except Exception:
        return is_port_reachable(config.coordinator.head_host, config.coordinator.ray_port, timeout)


def find_available_port(host: str, preferred: int, limit: int = 200) -> int:
    for port in range(preferred, preferred + limit):
        if is_port_available(host, port):
            return port
    raise RuntimeError(f"No available port found starting at {preferred}")


def ray_version() -> str | None:
    return bootstrap_ray_version(timeout=5)


def python_version() -> str | None:
    return bootstrap_python_version(timeout=5)


def has_ray() -> bool:
    return resolved_ray_executable() is not None


def has_worker_account(account: str) -> bool:
    if platform.system() == "Windows":
        result = subprocess.run(["net", "user", account], check=False, capture_output=True, text=True)
        return result.returncode == 0
    try:
        import pwd

        pwd.getpwnam(account)
        return True
    except KeyError:
        return False


def has_worker_launch_permission(account: str) -> bool:
    if not account or platform.system() == "Windows":
        return True
    if platform.system() != "Darwin":
        return True
    try:
        result = subprocess.run(["sudo", "-n", "-u", account, "--", "true"], check=False, capture_output=True, text=True, timeout=5)
    except (OSError, subprocess.TimeoutExpired):
        return False
    return result.returncode == 0


def container_runtime_status(runtime: str) -> tuple[bool, str]:
    binary = shutil.which(runtime)
    if not binary:
        return False, f"{runtime} is not installed or not on PATH"
    try:
        result = subprocess.run([runtime, "--version"], check=False, capture_output=True, text=True, timeout=5)
    except subprocess.TimeoutExpired:
        return False, f"{runtime} did not respond to --version"
    if result.returncode != 0:
        return False, (result.stderr or result.stdout or f"{runtime} failed").strip()
    return True, (result.stdout or result.stderr).strip()


def gpu_runtime_status(runtime: str) -> tuple[bool, str]:
    if runtime == "docker":
        cmd = ["docker", "info", "--format", "{{json .Runtimes}}"]
    else:
        cmd = [runtime, "info"]
    try:
        result = subprocess.run(cmd, check=False, capture_output=True, text=True, timeout=8)
    except (OSError, subprocess.TimeoutExpired):
        return False, "GPU runtime check could not run"
    text = (result.stdout + result.stderr).lower()
    if "nvidia" in text or "gpu" in text:
        return True, "GPU runtime appears configured"
    return False, "No NVIDIA/GPU runtime detected"


def idle_snapshot(cpu_threshold: int) -> bool:
    return psutil.cpu_percent(interval=0.2) <= cpu_threshold


def diagnostics(config: AppConfig) -> list[DiagnosticCheck]:
    checks: list[DiagnosticCheck] = []
    rv = ray_version()
    pyv = python_version()
    ray_ok = rv == PINNED_RAY_VERSION and pyv == PINNED_PYTHON
    checks.append(
        DiagnosticCheck(
            id="ray",
            label="Ray CLI",
            status="pass" if ray_ok else "fail",
            detail=(f"Ray {rv}, Python {pyv} at {resolved_ray_executable()}" if rv else "Ray CLI not found"),
            fix=f"Install the app-local pinned Ray runtime ({PINNED_RAY_VERSION}) with Python {PINNED_PYTHON} from Setup.",
        )
    )
    private = is_private_host(config.coordinator.head_host)
    checks.append(
        DiagnosticCheck(
            id="private_network",
            label="Private VLAN address",
            status="pass" if private else "fail",
            detail=f"Coordinator host {config.coordinator.head_host} {'is private' if private else 'is not private or cannot be resolved'}",
            fix="Use a private VLAN IP/DNS name; do not bind Ray to public campus or internet interfaces.",
        )
    )
    if config.app_mode.value == "node":
        reachable = is_coordinator_reachable(config)
        checks.append(
            DiagnosticCheck(
                id="ray_port",
                label="Coordinator reachability",
                status="pass" if reachable else "warn",
                detail=(
                    f"Coordinator is reachable at {config.coordinator.ray_address}"
                    if reachable
                    else f"Coordinator is not reachable at {config.coordinator.ray_address}"
                ),
                fix="Start the host in External workers mode or save the host machine's LAN IP address.",
            )
        )
    else:
        port_ok = is_port_available(config.coordinator.head_host, config.coordinator.ray_port)
        checks.append(
            DiagnosticCheck(
                id="ray_port",
                label="Ray head port",
                status="pass" if port_ok else "fail",
                detail=f"{config.coordinator.head_host}:{config.coordinator.ray_port} {'is available' if port_ok else 'is already in use'}",
                fix="Run Full machine setup to automatically select an available Ray head port.",
            )
        )
    account_ok = has_worker_account(config.privacy.worker_account)
    launch_ok = not account_ok or has_worker_launch_permission(config.privacy.worker_account)
    coordinator_mac = config.app_mode.value == "coordinator" and platform.system() == "Darwin"
    node_mac_missing_permission = config.app_mode.value == "node" and platform.system() == "Darwin" and account_ok and not launch_ok
    checks.append(
        DiagnosticCheck(
            id="worker_account",
            label="Dedicated worker account",
            status="fail" if node_mac_missing_permission else "pass" if account_ok else ("warn" if coordinator_mac else "fail"),
            detail=(
                f"Account {config.privacy.worker_account} exists, but RayLab needs administrator approval to launch worker processes"
                if node_mac_missing_permission
                else
                f"Account {config.privacy.worker_account} exists"
                if account_ok
                else "macOS Coordinator/UI mode can continue; production GPU worker isolation is Windows/Linux only"
                if coordinator_mac
                else f"Account {config.privacy.worker_account} does not exist"
            ),
            fix=(
                "Run Full machine setup and approve the macOS administrator prompt."
                if node_mac_missing_permission
                else "Use a Linux/Windows worker node for GPU sharing."
                if coordinator_mac and not account_ok
                else "Create the raylab-worker account using the OS setup guide before enabling sharing."
            ),
        )
    )
    runtime_ok, runtime_detail = container_runtime_status(config.privacy.container_runtime)
    checks.append(
        DiagnosticCheck(
            id="container_runtime",
            label="Container runtime",
            status="pass" if runtime_ok else "fail",
            detail=runtime_detail,
            fix="Install Docker/Podman and configure it for the dedicated worker account.",
        )
    )
    gpu_ok, gpu_detail = gpu_runtime_status(config.privacy.container_runtime) if runtime_ok else (False, "Skipped")
    checks.append(
        DiagnosticCheck(
            id="gpu_container_runtime",
            label="GPU container runtime",
            status="pass" if gpu_ok else "warn",
            detail=gpu_detail,
            fix="Install NVIDIA Container Toolkit or equivalent GPU support for the selected runtime.",
        )
    )
    store_ok = bool(config.object_store.endpoint_url and config.object_store.bucket)
    checks.append(
        DiagnosticCheck(
            id="object_store",
            label="Lab object store",
            status="pass" if store_ok else "warn",
            detail="Configured" if store_ok else "Endpoint or bucket is missing",
            fix="Set the S3-compatible endpoint and bucket used for datasets and model weights.",
        )
    )
    return checks


@dataclass(frozen=True)
class SetupCommand:
    os_name: str
    command: str
    note: str


def worker_account_setup_commands(account: str = "raylab-worker") -> list[SetupCommand]:
    return [
        SetupCommand("Linux", f"sudo useradd --system --create-home --shell /usr/sbin/nologin {account}", "Create an unprivileged system user for Ray workers."),
        SetupCommand("Linux", f"sudo mkdir -p /var/lib/{account}/jobs && sudo chown -R {account}:{account} /var/lib/{account}", "Create the job working directory."),
        SetupCommand("Windows", f"net user {account} * /add", "Create a local user with a generated password entered by an administrator."),
        SetupCommand("Windows", f"mkdir C:\\RayLab\\jobs", "Create a controlled job working directory."),
    ]

File: electron/src/diagnostics.js
'use strict';

const net = require('net');
const dns = require('dns');
const { spawnSync } = require('child_process');
const { rayVersion, pythonVersion, rayCommand, PINNED_RAY_VERSION, PINNED_PYTHON, resolvedRayExecutable, which } = require('./bootstrap');

// ─── Private-host check ───────────────────────────────────────────────────────

async function isPrivateHost(host) {
  if (!host) return false;
  if (host === '127.0.0.1' || host === 'localhost' || host === '::1') return true;
  try {
    const { address } = await dns.promises.lookup(host);
    return _isPrivateIp(address);
  } catch (_) {
    return false;
  }
}

function _isPrivateIp(ip) {
  if (ip.startsWith('127.') || ip === '::1' || ip === 'localhost') return true;
  if (ip.startsWith('10.')) return true;
  if (ip.startsWith('192.168.')) return true;
  const match = ip.match(/^172\.(\d+)\./);
  if (match && parseInt(match[1]) >= 16 && parseInt(match[1]) <= 31) return true;
  return false;
}

// ─── Port helpers ─────────────────────────────────────────────────────────────

function isPortAvailable(host, port) {
  return new Promise((resolve) => {
    const server = net.createServer();
    server.once('error', () => resolve(false));
    server.once('listening', () => { server.close(); resolve(true); });
    server.listen(port, host === 'localhost' ? '127.0.0.1' : host);
  });
}

function isPortReachable(host, port, timeoutMs = 1000) {
  return new Promise((resolve) => {
    const sock = new net.Socket();
    sock.setTimeout(timeoutMs);
    sock.once('connect', () => { sock.destroy(); resolve(true); });
    sock.once('error', () => { sock.destroy(); resolve(false); });
    sock.once('timeout', () => { sock.destroy(); resolve(false); });
    sock.connect(port, host);
  });
}

async function findAvailablePort(host, preferred, limit = 200) {
  for (let port = preferred; port < preferred + limit; port++) {
    if (await isPortAvailable(host, port)) return port;
  }
  throw new Error(`No available port found in range ${preferred}–${preferred + limit - 1}`);
}

// ─── Coordinator reachability ─────────────────────────────────────────────────

async function isCoordinatorReachable(config) {
  const host = config.coordinator.head_host;
  const port = config.coordinator.ray_port;
  if (await isPortReachable(host, port, 1000)) return true;
  // nc CLI fallback (macOS)
  if (process.platform === 'darwin') {
    const nc = '/usr/bin/nc';
    try {
      const result = spawnSync(nc, ['-z', '-G', '2', host, String(port)], { timeout: 3000 });
      if (result.status === 0) return true;
    } catch (_) {}
  }
  return false;
}

// ─── Worker account checks ────────────────────────────────────────────────────

function hasWorkerAccount(account) {
  if (!account) return false;
  if (process.platform === 'win32') {
    try {
      const r = spawnSync('net', ['user', account], { timeout: 5000, encoding: 'utf8' });
      return r.status === 0;
    } catch (_) { return false; }
  }
  try {
    const r = spawnSync('id', [account], { timeout: 5000, encoding: 'utf8' });
    return r.status === 0;
  } catch (_) { return false; }
}

function hasWorkerLaunchPermission(account) {
  if (!account) return true;
  if (process.platform !== 'darwin') return true;
  try {
    const r = spawnSync('sudo', ['-n', '-u', account, '--', 'true'], { timeout: 5000 });
    return r.status === 0;
  } catch (_) { return false; }
}

// ─── Container runtime checks ─────────────────────────────────────────────────

function containerRuntimeStatus(runtime) {
  const name = 'docker';
  const bin = dockerExecutable();
  if (!bin) return { ok: false, detail: 'Docker is not installed or not on PATH' };
  try {
    const r = spawnSync(bin, ['--version'], { timeout: 5000, encoding: 'utf8' });
    const detail = (r.stdout || r.stderr || '').trim();
    return { ok: r.status === 0, detail: detail || `${name} found` };
  } catch (_) {
    return { ok: false, detail: 'Docker is not available' };
  }
}

function dockerExecutable() {
  const fromPath = which('docker');
  if (fromPath) return fromPath;
  if (process.platform === 'win32') {
    const fs = require('fs');
    const candidates = [
      'C:\\Program Files\\Docker\\Docker\\resources\\bin\\docker.exe',
      'C:\\Program Files\\Docker\\Docker\\resources\\docker.exe',
    ];
    return candidates.find((p) => fs.existsSync(p)) || null;
  }
  return null;
}

function gpuRuntimeStatus(runtime) {
  try {
    const docker = dockerExecutable() || 'docker';
    const result = spawnSync(docker, ['info', '--format', '{{json .Runtimes}}'], { timeout: 8000, encoding: 'utf8' });
    const text = ((result.stdout || '') + (result.stderr || '')).toLowerCase();
    if (text.includes('nvidia') || text.includes('gpu')) {
      return { ok: true, detail: 'GPU runtime appears configured' };
    }
    return { ok: false, detail: 'No NVIDIA/GPU runtime detected' };
  } catch (_) {
    return { ok: false, detail: 'Container runtime check failed' };
  }
}

// ─── 7 diagnostic checks ─────────────────────────────────────────────────────

async function diagnostics(config) {
  const checks = [];
  const mode = config.app_mode;
  const coord = config.coordinator;
  const privacy = config.privacy;
  const store = config.object_store;

  // 1 — Ray CLI
  const rayExe = resolvedRayExecutable();
  const rv = rayExe ? rayVersion(rayExe) : null;
  const pv = pythonVersion();
  const rayOk = rv === PINNED_RAY_VERSION && pv === PINNED_PYTHON;
  checks.push({
    id: 'ray',
    label: 'Ray CLI',
    status: rayOk ? 'pass' : 'fail',
    detail: rayOk
      ? `Ray ${rv}, Python ${pv} at ${rayExe}`
      : (rv ? `Ray ${rv} found but expected ${PINNED_RAY_VERSION}` : 'Ray CLI not found'),
    fix: rayOk ? null : `Install the app-local pinned Ray runtime (${PINNED_RAY_VERSION}) with Python ${PINNED_PYTHON} from Setup.`,
  });

  // 2 — Private network
  const privateOk = await isPrivateHost(coord.head_host);
  checks.push({
    id: 'private_network',
    label: 'Private VLAN address',
    status: privateOk ? 'pass' : 'fail',
    detail: privateOk
      ? `Coordinator host ${coord.head_host} is private`
      : `${coord.head_host} is not private or cannot be resolved`,
    fix: privateOk ? null : 'Use a private VLAN IP/DNS name; do not bind Ray to public campus or internet interfaces.',
  });

  // 3 — Ray port
  if (mode === 'node') {
    const reachable = await isCoordinatorReachable(config);
    checks.push({
      id: 'ray_port',
      label: 'Coordinator reachability',
      status: reachable ? 'pass' : 'warn',
      detail: reachable
        ? `Coordinator reachable at ${coord.head_host}:${coord.ray_port}`
        : `Cannot reach ${coord.head_host}:${coord.ray_port}`,
      fix: reachable ? null : 'Start the host in External workers mode or save the host machine\'s LAN IP address.',
    });
  } else {
    const portFree = await isPortAvailable(coord.head_host, coord.ray_port);
    checks.push({
      id: 'ray_port',
      label: 'Ray head port',
      status: portFree ? 'pass' : 'fail',
      detail: portFree
        ? `Port ${coord.ray_port} is available on ${coord.head_host}`
        : `Port ${coord.ray_port} on ${coord.head_host} is already in use`,
      fix: portFree ? null : 'Run Full machine setup to automatically select an available Ray head port.',
    });
  }

  // 4 — Worker account
  const account = privacy.worker_account || 'raylab-worker';
  const accountExists = hasWorkerAccount(account);
  let workerStatus, workerDetail, workerFix;
  if (process.platform === 'darwin' && mode === 'node' && accountExists && !hasWorkerLaunchPermission(account)) {
    workerStatus = 'fail';
    workerDetail = `Account ${account} exists but sudo launch permission not yet granted`;
    workerFix = 'Run Full machine setup and approve the macOS administrator prompt.';
  } else if (accountExists) {
    workerStatus = 'pass';
    workerDetail = `Dedicated account ${account} is ready`;
    workerFix = null;
  } else if (process.platform === 'darwin' && mode === 'coordinator') {
    workerStatus = 'warn';
    workerDetail = 'macOS Coordinator/UI mode can continue; production GPU worker isolation is Windows/Linux only';
    workerFix = null;
  } else {
    workerStatus = 'fail';
    workerDetail = `Account ${account} does not exist`;
    workerFix = 'Use Create account to create the dedicated worker account with administrator approval.';
  }
  checks.push({ id: 'worker_account', label: 'Dedicated worker account', status: workerStatus, detail: workerDetail, fix: workerFix });

  // 5 — Container runtime
  const runtime = 'docker';
  const containerStatus = containerRuntimeStatus(runtime);
  checks.push({
    id: 'container_runtime',
    label: 'Container runtime',
    status: containerStatus.ok ? 'pass' : 'fail',
    detail: containerStatus.detail,
    fix: containerStatus.ok ? null : 'Use Install Docker to install Docker Desktop, then restart Docker if Windows asks.',
  });

  // 6 — GPU container runtime
  if (!containerStatus.ok) {
    checks.push({ id: 'gpu_container_runtime', label: 'GPU container runtime', status: 'warn', detail: 'Skipped (container runtime not available)', fix: null });
  } else {
    const gpuStatus = gpuRuntimeStatus(runtime);
    checks.push({
      id: 'gpu_container_runtime',
      label: 'GPU container runtime',
      status: gpuStatus.ok ? 'pass' : 'warn',
      detail: gpuStatus.detail,
      fix: gpuStatus.ok ? null : 'Install NVIDIA Container Toolkit or equivalent GPU support for the selected runtime.',
    });
  }

  // 7 — Object store
  const storeOk = !!(store.endpoint_url && store.bucket);
  checks.push({
    id: 'object_store',
    label: 'Lab object store',
    status: storeOk ? 'pass' : 'warn',
    detail: storeOk ? `${store.endpoint_url} / ${store.bucket}` : 'Object store is not configured yet',
    fix: storeOk ? null : 'Set endpoint and bucket before submitting real LLM/data jobs.',
  });

  return checks;
}

module.exports = {
  diagnostics,
  isPrivateHost,
  isPortAvailable,
  isPortReachable,
  findAvailablePort,
  isCoordinatorReachable,
  hasWorkerAccount,
  hasWorkerLaunchPermission,
  containerRuntimeStatus,
  gpuRuntimeStatus,
  dockerExecutable,
};
